src/wallet/wallet.cpp

```diff
@@ -1364,9 +1364,13 @@ CAmount CWallet::GetDebit(const CTransaction& tx, const isminefilter& filter) co
 bool CWallet::IsHDEnabled() const
 {
     // All Active ScriptPubKeyMans must be HD for this to be true
-    bool result = true;
+    bool result = false;
     for (const auto& spk_man : GetActiveScriptPubKeyMans()) {
-        result &= spk_man->IsHDEnabled();
+        if (spk_man->IsHDEnabled()) {
+            result = true;
+        } else {
+            return false;
+        }
```

Comment:
Or, even better:
```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1364,11 +1364,11 @@ CAmount CWallet::GetDebit(const CTransaction& tx, const isminefilter& filter) co
 bool CWallet::IsHDEnabled() const
 {
     // All Active ScriptPubKeyMans must be HD for this to be true
-    bool result = true;
+    if (GetActiveScriptPubKeyMans().empty()) return false;
     for (const auto& spk_man : GetActiveScriptPubKeyMans()) {
-        result &= spk_man->IsHDEnabled();
+        if (!spk_man->IsHDEnabled()) return false;
     }
-    return result;
+    return true;
 }
 
 bool CWallet::CanGetAddresses(bool internal) const
```
?

Reply:
I thought too about using a check if keys are empty, but then was not sure if empty will work correctly.
So the second suggestion ***might*** work, the first not, since we will never enter the loop when wallet is empty, and subsequently return true, when indeed false would be correct.

I saw btw the discussion https://github.com/bitcoin-core/gui/pull/403 and found that the the statement the pr works and the behavior after refraktor has not changed quite revealing, i tested too that PR and saw that it would not change this vital aspect, but i guess it is not only a GUI issue and to some extent this PR is also a hint to the implications of constructions in the form of bare
`for (const auto& x : X()) ...`. when we can no be sure about the size of x and make vital decisions only in the loop.

Reply:
> ... the first not, since we will never enter the loop when wallet is empty, and subsequently return true, when indeed false would be correct.

I don't think this is correct description of the suggested change. Both lines are within a loop body, and there is no `return true;` statement at all.

Reply:
yup :+1: , my fault , the init value is still false, your suggestion is better...